modules/packages/Sort.chpl

```diff
@@ -18,27 +18,178 @@
  */
 
 /*
-   The `Sort` module is designed to support standard sorting
-   routines.  The current interface is minimal and should be expected to
-   grow and evolve over time. Each of the following functions accepts an
-   optional boolean argument, `reverse`, which is `false` by
-   default. If `true`, then the sort function will order in reverse.
+
+This module supports a variety of standard sorting routines on 1D arrays.
+
+The current interface is minimal and should be expected to
+grow and evolve over time.
+
+Comparators
+-----------
+
+Comparators allow sorting data by a mechanism other than the
+default comparison of array elements. To use a comparator, define a
+record with either a ``key(a)`` or ``compare(a, b)`` method, and pass an
+instance of that record to the sort function.
+
+Most sort routines support comparators, which is denoted by their function
+signature.
+
+The ``key(a)`` method accepts one argument with the array ``eltType`` and
+returns a value that supports all of comparison operations: ``<``, ``<=``,
+``>``, ``>=``, or ``==``.
+
+The default key method would look like this:
+
+.. code-block:: chapel
+
+  proc default.key(a) {
+    return a;
+  }
+
+
+As an example, if we want to sort an array by the absolute values of its
+elements, we can use the following comparator record with a key method:
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record comparator { }
+
+  // key method maps an element to the value to be used for comparison
+  proc comparator.key(a) { return abs(a); }
+
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+
+The ``compare(a, b)`` method accepts 2 arguments of the data ``eltType`` and
+returns an integer between ``1`` if ``a > b``, ``0`` if ``a == b``, and ``-1`` if ``a < b``.
+
+The default compare method would look like this:
+
+.. code-block:: chapel
+
+    proc default.compare(a, b) {
+      if a > b {
+        return 1;
+      } else if a < b {
+        return -1;
+      } else {
+        return 0;
+      }
+    }
+
+
+The absolute value comparison example from above can alternatively be
+implemented with a compare method:
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc comparator.compare(a, b) {
+    if abs(a) > abs(b) {
+      return 1;
+    } else if abs(a) < abs(b) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+If both methods are implemented on the record passed as the comparator, the
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
```

Comment:
Hmm... interesting. Could you mark this as "current behavior" somehow? Once we add radix sort (which I think is ... key ... heh) it will not be able to use compare at all.

E.g.

In the current implementation, if both...